### PR TITLE
Create ObjectOwnedBase class

### PR DESCRIPTION
**Pull request by blowekamp**
Refactor components from the Command object to be reused by other objects.

### CODE REVIEW

**Comment by blowekamp** [comment]
@dave3d The spell checker is reporting a false positive for "copiable". How can this bee addressed?

**Comment by dave3d** [comment]
> @dave3d The spell checker is reporting a false positive for "copiable". How can this bee addressed?

I can add the word to the dictionary.

**Review by dave3d — APPROVED** [review]

File: Code/Common/include/sitkObjectOwnedBase.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  SetName(const std::string & name);
+
+protected:
+
+  /** \brief Copy Constructor
+   *
+   * Only copiable if o does no have OwnedByObjects enabled. If OwnedByObjects is enabled in o, than an exception will
+   * be thrown.
+   *
+   * The object callbacks are copied the the new object.
```
Comment (zivy) [inline Code/Common/include/sitkObjectOwnedBase.h:75]: "the the" -> "to the"

File: Code/Common/include/sitkObjectOwnedBase.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+  /** Get object name. */
+  virtual void
+  SetName(const std::string & name);
+
+protected:
+
+  /** \brief Copy Constructor
+   *
+   * Only copiable if o does no have OwnedByObjects enabled. If OwnedByObjects is enabled in o, than an exception will
```
Comment (zivy) [inline Code/Common/include/sitkObjectOwnedBase.h:72]: "does no" -> "does not"

File: Code/Common/include/sitkObjectOwnedBase.h (hunk 1 above)
Comment (zivy) [inline Code/Common/include/sitkObjectOwnedBase.h:75]: "the the" -> "to the"

File: Code/Common/include/sitkObjectOwnedBase.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+   *
+   * The object callbacks are copied the the new object.
+   */
+  ObjectOwnedBase(const ObjectOwnedBase & o);
+
+#ifndef SWIG
+  /** \brief Register an Object and a method to call on this object's destructor.
+   *
+   * @param o
+   * @param onDelete A function object to be call
```
Comment (zivy) [inline Code/Common/include/sitkObjectOwnedBase.h:83]: "to be call" -> "to be called"

**Review by zivy — APPROVED** [review]
minor grammar

File: Code/Common/include/sitkObjectOwnedBase.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+ * Optionally, "OwnedByObjects" property can be enabled so that when the last referenced object is removed this object
+ * will be deleted.
+ *
+ *
+ */
+class SITKCommon_EXPORT ObjectOwnedBase
+: public NonCopyable
+{
+public:
+  virtual ~ObjectOwnedBase() = 0;
```
Comment (blowekamp) [inline Code/Common/include/sitkObjectOwnedBase.h:51]: After an excessive amount of trial, error, and simplifying the R SWIG error, it appears that the `= 0` is causing problems with R SWIG. Perhaps this was causing issue with the Python wrapping I encountered earlier too.

**Pull request merged**